### DISCUSSION TITLE
How does the algorithm work?

### DISCUSSION THREAD

**Discussion by IllustratedMan-code** [Schrodinger-Hat/ImageGoNord#130] [Ideas]
I think it would be great to have some documentation on the pros and cons of the current conversion algorithm. From the Python package, it seems like `Image.quantize()` method from PIL is where the actual conversion is happening, though I could be wrong. Is there any benefit to this method over the naive (Euclidian) technique in this [article](https://en.wikipedia.org/wiki/Color_difference)?

[I implemented a version of this in C++ with opencv](https://github.com/IllustratedMan-code/Image-Factory) and it seems like it *might* be faster, though I only did a few tests against the ImageGoNord [website](https://ign.schrodinger-hat.it/getting-started). My method is extremely basic, with essentially 0 optimizations. Opencv does use parallel CPU operations so it's still pretty fast.

I would love to hear some thoughts on possible improvements to the algorithm!

**Comment by TheJoin95** [comment]
We are actually using three different kind of modulable algorithm: quantize, pixel by pixel, avg area.
The first one is just a quantize of the palettes: you've got 500 colors in an image, you set a palette of just 16, it will try to match the nearest color onto those 16 available.

The pixel by pixel is again using the color difference, but it taking half life to convert an image.

The last one is the one that we like the most: it analyze an area of pixels (determineted from the input) and compare it to the previous one, making an average and taking the nearest color of the new palette. It's making this evalutation each time the "matrix of the avg" is moving towards the image.

It is not optimized, as you can see but the results are quite good, on our opinion (without taking in consideration of the time)

**Reply by IllustratedMan-code** [reply]
I don't think I quite understand the avg area method. Are you just blurring the image before processing ([linear box filter](https://docs.opencv.org/3.4/d4/dbd/tutorial_filter_2d.html)), then [blending](https://docs.opencv.org/3.4/d5/dc4/tutorial_adding_images.html) the two images together?

Here is what I was able to come up with:

blend ratio (0.8/0.2) | blend ratio (0.5/0.5) | blend ratio (0.8/0.2) | image go nord (slider to the right)
:-----:|:------:|:-------:|:------:
![h](https://user-images.githubusercontent.com/60514384/203146849-568515ce-f03f-46e4-9f56-b45060588769.jpg) | ![h](https://user-images.githubusercontent.com/60514384/203146950-b36e9f33-58cc-45da-888c-7ca0ec75ca66.jpg) | ![h](https://user-images.githubusercontent.com/60514384/203147059-080ed066-b889-4c2a-b028-559153e38d86.jpg) | ![test](https://user-images.githubusercontent.com/60514384/203147138-d77adb18-ef63-47fd-8ac3-12aeb34690d5.png) 


I assume something else is going on because I can't quite match the image go nord output. Am I misunderstanding? I quite like this blend method though, it makes the colorization a lot more natural looking.
